encoders/firebase-decoders-json/src/test/java/com/google/firebase/decoders/TypeTokenTest.java

```diff
@@ -0,0 +1,151 @@
+    assertThat(componentType).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<List<String>> componentClassType = (TypeToken.ClassToken<List<String>>) componentType;
+    assertThat(componentClassType.getRawType()).isEqualTo(List.class);
+    TypeToken<String> argumentType = ((TypeToken.ClassToken<List<String>>) componentType).getTypeArguments().at(0);
+    assertThat(argumentType).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<String> argumentClassType = (TypeToken.ClassToken<String>) argumentType;
+    assertThat(argumentClassType.getRawType()).isEqualTo(String.class);
+  }
+
+  //Plain Class Type
+  @Test
+  public void plainClassTypeWithSafe_rawTypeIsCorrectlyCaptured() {
+    TypeToken<Foo> typeToken = TypeToken.of(new Safe<Foo>() {});
+    assertThat(typeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<Foo> typeClassToken = (TypeToken.ClassToken<Foo>) typeToken;
+    assertThat(typeClassToken.getRawType()).isEqualTo(Foo.class);
+  }
+
+  @Test
+  public void plainClassTypeWithoutSafe_rawTypeIsCorrectlyCaptured() {
+    TypeToken<Foo> typeToken = TypeToken.of(Foo.class);
+    assertThat(typeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<Foo> typeClassToken = (TypeToken.ClassToken<Foo>) typeToken;
+    assertThat(typeClassToken.getRawType()).isEqualTo(Foo.class);
+  }
+
+  //Generic Type
+  @Test
+  public void genericType_actualTypeParametersAreCorrectlyCaptured(){
+    TypeToken<Map<String, Foo>> mapTypeToken = TypeToken.of(new Safe<Map<String, Foo>>() {});
+    assertThat(mapTypeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<Map<String, Foo>> mapClassToken = (TypeToken.ClassToken<Map<String, Foo>>) mapTypeToken;
+    TypeTokenContainer typeTokenContainer = mapClassToken.getTypeArguments();
+    TypeToken<String> firstArgumentToken = typeTokenContainer.at(0);
+    TypeToken<Foo> secondArgumentTypeToken = typeTokenContainer.at(1);
+    assertThat(firstArgumentToken).isInstanceOf(TypeToken.ClassToken.class);
+    assertThat(secondArgumentTypeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<String> firstArgumentClassToken = (TypeToken.ClassToken<String>) firstArgumentToken;
+    TypeToken.ClassToken<Foo> secondArgumentClassToken = (TypeToken.ClassToken<Foo>) secondArgumentTypeToken;
+    assertThat(firstArgumentClassToken.getRawType()).isEqualTo(String.class);
+    assertThat(secondArgumentClassToken.getRawType()).isEqualTo(Foo.class);
+  }
+
+  @Test
+  public void nestedGenericType_actualTypeParametersAreCorrectlyCaptured(){
+    TypeToken<List<List<String>>> typeToken = TypeToken.of(new Safe<List<List<String>>>(){});
+    assertThat(typeToken).isInstanceOf(TypeToken.ClassToken.class);
+    TypeToken.ClassToken<List<List<String>>> typeClassToken = (TypeToken.ClassToken<List<List<String>>>) typeToken;
+    TypeToken<List<String>> componentToken = typeClassToken.getTypeArguments().at(0);
```

Comment:
How does this test work? afaict typeArguments is `null` in your implementation.

Reply:
`typeArguments` is `null` only if the `ClassToken` contains non-generic class type. Otherwise, it will be able to get its actual type parameters.
This test makes sure every actual type parameters(as well as the inner component type) can be captured correctly.

Reply:
instead of returning null, can we return an empty `TypeTokenContainer` instead? Also might be worth making `TypeTokenContainer` a class instead of an interface.

On second thought it's fine this way for this PR, but please address this in a follow up pr.